docs/repo/crates/discv4.md

```diff
@@ -0,0 +1,316 @@
+
+        let udp = Arc::clone(&socket);
+        tasks.spawn(async move { receive_loop(udp, ingress_tx, local_node_record.id).await });
+
+        let udp = Arc::clone(&socket);
+        tasks.spawn(async move { send_loop(udp, egress_rx).await });
+
+        let kbuckets = KBucketsTable::new(
+            NodeKey::from(&local_node_record).into(),
+            Duration::from_secs(60),
+            MAX_NODES_PER_BUCKET,
+            None,
+            None,
+        );
+
+        let self_lookup_interval = tokio::time::interval(config.lookup_interval);
+
+        // Wait `ping_interval` and then start pinging every `ping_interval`
+        let ping_interval = tokio::time::interval_at(
+            tokio::time::Instant::now() + config.ping_interval,
+            config.ping_interval,
+        );
+
+        let evict_expired_requests_interval = tokio::time::interval_at(
+            tokio::time::Instant::now() + config.request_timeout,
+            config.request_timeout,
+        );
+
+        //--snip--
+    }
+//--snip--
+}
+```
+
+First, new channels are initialized to handle incoming and outgoing traffic related to node discovery. New tasks are then spawned to handle the `receive_loop()` and `send_loop()` functions, which perpetually send incoming disc4 traffic to the `ingress_rx` and outgoing discv4 traffic to the `egress_rx`. Following this, a [`KBucketsTable`]() is initialized to keep track of the node's neighbors throughout the network. If you are unfamiliar with k-buckets, feel free to [follow this link to learn more](https://en.wikipedia.org/wiki/Kademlia#Fixed-size_routing_tables). Following this, the `self_lookup_interval`, `ping_interval` and `evict_expired_requests_interval` are initialized which determine the intervals between self lookups, outgoing ping packets and the time elapsed to remove inactive neighbors from the node's records.
+
+Once the `Discv4Service::new()` function completes, allowing the `Discv4::bind()` function to complete as well, the `discv4_service` is then spawned into its own task and the `Discovery::new()` function returns an new instance of `Discovery`. The newly created `Discovery` type is passed into the `NetworkState::new()` function along with a few other arguments to initialize the network state. The network state is added to the `NetworkManager` where the `NetworkManager` is then spawned into its own task as well.
+
+
+
+## Polling the Discv4Service and Discovery Events
+In Rust, the owner of a [`Future`](https://doc.rust-lang.org/std/future/trait.Future.html#) is responsible for advancing the computation by polling the future. This is done by calling `Future::poll`.
+
+Lets take a detailed look at how `Discv4Service::poll` works under the hood. This function has many moving parts, so we will break it up into smaller sections. If you would like to check out the function in it's entirety, you can [click here]().
+
+//TODO: highlight where the future is actually being spawned
```

Comment:
You pretty much already do - it's spawned by the `Discv4Service::new` function which is called in `Discv4::bind`